rust/src/tx_builder.rs

```diff
@@ -154,13 +154,113 @@ pub enum CoinSelectionStrategyCIP2 {
 
 #[wasm_bindgen]
 #[derive(Clone, Debug)]
-pub struct TransactionBuilder {
-    coins_per_utxo_word: BigNum,
-    pool_deposit: BigNum,
-    key_deposit: BigNum,
-    max_value_size: u32,
-    max_tx_size: u32,
+pub struct TransactionBuilderConfig {
     fee_algo: fees::LinearFee,
+    pool_deposit: BigNum,      // protocol parameter
+    key_deposit: BigNum,       // protocol parameter
+    max_value_size: u32,       // protocol parameter
+    max_tx_size: u32,          // protocol parameter
+    coins_per_utxo_word: Coin, // protocol parameter
+    prefer_pure_change: bool,
+}
+
+#[wasm_bindgen]
+#[derive(Clone, Debug)]
+pub struct TransactionBuilderConfigBuilder {
+    fee_algo: Option<fees::LinearFee>,
+    pool_deposit: Option<BigNum>,      // protocol parameter
+    key_deposit: Option<BigNum>,       // protocol parameter
+    max_value_size: Option<u32>,       // protocol parameter
+    max_tx_size: Option<u32>,          // protocol parameter
+    coins_per_utxo_word: Option<Coin>, // protocol parameter
+    prefer_pure_change: bool,
+}
+
+impl Default for TransactionBuilderConfigBuilder {
+    fn default() -> Self {
+        Self {
+            fee_algo: None,
+            pool_deposit: None,
+            key_deposit: None,
+            max_value_size: None,
+            max_tx_size: None,
+            coins_per_utxo_word: None,
+            prefer_pure_change: false,
+        }
+    }
+}
+
+impl TransactionBuilderConfigBuilder {
+    pub fn fee_algo(&mut self, fee_algo: fees::LinearFee) -> &mut Self {
+        self.fee_algo = Some(fee_algo);
+        self
+    }
+
+    pub fn pool_deposit(&mut self, pool_deposit: BigNum) -> &mut Self {
+        self.pool_deposit = Some(pool_deposit);
+        self
+    }
+
+    pub fn key_deposit(&mut self, key_deposit: BigNum) -> &mut Self {
+        self.key_deposit = Some(key_deposit);
+        self
+    }
+
+    pub fn max_value_size(&mut self, max_value_size: u32) -> &mut Self {
+        self.max_value_size = Some(max_value_size);
+        self
+    }
+
+    pub fn max_tx_size(&mut self, max_tx_size: u32) -> &mut Self {
+        self.max_tx_size = Some(max_tx_size);
+        self
+    }
+
+    pub fn coins_per_utxo_word(&mut self, coins_per_utxo_word: Coin) -> &mut Self {
+        self.coins_per_utxo_word = Some(coins_per_utxo_word);
+        self
+    }
+
+    pub fn prefer_pure_change(&mut self, prefer_pure_change: bool) -> &mut Self {
+        self.prefer_pure_change = prefer_pure_change;
+        self
+    }
+
+    pub fn build(&self) -> Result<TransactionBuilderConfig, JsError> {
+        Ok(TransactionBuilderConfig {
+            fee_algo: match self.fee_algo {
+                Some(ref fee_algo) => fee_algo.clone(),
+                None => return Err(JsError::from_str("uninitialized field: fee_algo")),
+            },
```

Comment:
We can probably remove a bit of duplication with like a macro:
```
Ok(TransactionBuilderConfig {
  fee_algo: builder_get!(self, fee_algo),
  ...
})
```

```
macro_rules! builder_get {
    ($name:ident) => {
        match self.$name {
            Some(ref $name) => $name.clone(),
            None => return Err(JsError::from_str("unitialized field: $name"))
        }
    }
}
```

or even just a function that takes an Option<T> and the field name and returns a Result<T, JsError>

Reply:
done in the last commit